File: app/src/main/java/com/zingkg/arkhamhorrorassistant/app/fragments/ReckoningFragment.java
package com.zingkg.arkhamhorrorassistant.app.fragments;

import android.os.Bundle;
import android.text.Html;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.zingkg.arkhamhorrorassistant.app.R;
import com.zingkg.arkhamhorrorassistant.xml.Reckoning;

/**
 * This fragment represents the text of a Reckoning card.
 */
public class ReckoningFragment extends CardFragment {
    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Override
    public View onCreateView(
        LayoutInflater inflater,
        ViewGroup container,
        Bundle savedInstanceState
    ) {
        super.onCreateView(inflater, container, savedInstanceState);
        return inflater.inflate(R.layout.fragment_reckoning, container, false);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        Reckoning card = importReckoning(getArguments());
        if (card.entry.isEmpty()) {
            ((LinearLayout) view.findViewById(R.id.reckoning_layout)).setGravity(
                Gravity.CENTER_HORIZONTAL
            );
            view.findViewById(R.id.reckoning_divider).setVisibility(View.INVISIBLE);
        } else {
            ((TextView) view.findViewById(R.id.reckoning_entry)).setText(Html.fromHtml(card.entry));
        }
        ((TextView) view.findViewById(R.id.reckoning_title)).setText(Html.fromHtml(card.title));
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.menu_decks, menu);
    }

    private static final String titleKey = "title";
    private static final String entryKey = "entry";
    private static final String expansionSetKey = "expansionSetKey";

    public static Bundle exportReckoning(Reckoning reckoning) {
        Bundle bundle = new Bundle();
        bundle.putString(titleKey, reckoning.title);
        bundle.putString(entryKey, reckoning.entry);
        bundle.putString(expansionSetKey, reckoning.expansionSet);
        return bundle;
    }

    public static Reckoning importReckoning(Bundle bundle) {
        return new Reckoning(
            bundle.getString(titleKey, ""),
            bundle.getString(entryKey, ""),
            bundle.getString(expansionSetKey, "")
        );
    }
}

File: app/src/main/kotlin/com/zingkg/arkhamhorrorboard/secondedition/app/fragments/ReckoningFragment.kt
package com.zingkg.arkhamhorrorboard.secondedition.app.fragments

import android.os.Bundle
import android.text.Html
import android.view.Gravity
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView

import com.zingkg.arkhamhorrorboard.secondedition.app.R
import com.zingkg.arkhamhorrorboard.secondedition.card.Cards
import com.zingkg.arkhamhorrorboard.secondedition.card.Reckoning

class ReckoningFragment : CardFragment() {
  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setHasOptionsMenu(true)
  }

  override fun onCreateView(
    inflater: LayoutInflater?,
    container: ViewGroup?,
    savedInstanceState: Bundle?
  ): View? {
    super.onCreateView(inflater, container, savedInstanceState)
    return inflater?.inflate(R.layout.fragment_reckoning, container, false)
  }

  override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
    super.onViewCreated(view, savedInstanceState)
    val (title, entry, expansionSet) = importReckoning(arguments)
    if (entry.isEmpty()) {
      view?.findViewById<LinearLayout>(R.id.reckoning_layout)?.gravity = Gravity.CENTER_HORIZONTAL
      view?.findViewById<View>(R.id.reckoning_divider)?.visibility = View.INVISIBLE
      view?.findViewById<View>(R.id.reckoning_expansion_set_divider)?.visibility = View.INVISIBLE
    } else {
      view?.findViewById<TextView>(R.id.reckoning_entry)?.text = Html.fromHtml(entry)
      view?.findViewById<TextView>(R.id.reckoning_expansion_set)?.text = Html.fromHtml(expansionSet.shortName)
    }
    view?.findViewById<TextView>(R.id.reckoning_title)?.text = Html.fromHtml(title)
  }

  override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
    super.onCreateOptionsMenu(menu, inflater)
    inflater?.inflate(R.menu.menu_decks, menu)
  }

  companion object {
    private const val TITLE_KEY = "title"
    private const val ENTRY_KEY = "entry"
    private const val EXPANSION_SET_KEY = "expansionSetKey"

    fun exportReckoning(reckoning: Reckoning): Bundle {
      val bundle = Bundle()
      bundle.putString(TITLE_KEY, reckoning.title)
      bundle.putString(ENTRY_KEY, reckoning.entry)
      bundle.putString(EXPANSION_SET_KEY, reckoning.expansionSet.toString())
      return bundle
    }

    fun importReckoning(bundle: Bundle): Reckoning {
      return Reckoning(
        bundle.getString(TITLE_KEY, ""),
        bundle.getString(ENTRY_KEY, ""),
        Cards.ExpansionSet.valueOf(bundle.getString(EXPANSION_SET_KEY, Cards.ExpansionSet.BASE.name))
      )
    }
  }
}
